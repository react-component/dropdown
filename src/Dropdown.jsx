import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ReactDOM from 'react-dom';
import Trigger from 'rc-trigger';
import contains from 'rc-util/lib/Dom/contains';
import addEventListener from 'rc-util/lib/Dom/addEventListener';
import placements from './placements';

export default class Dropdown extends Component {
  static propTypes = {
    minOverlayWidthMatchTrigger: PropTypes.bool,
    onVisibleChange: PropTypes.func,
    onOverlayClick: PropTypes.func,
    prefixCls: PropTypes.string,
    children: PropTypes.any,
    transitionName: PropTypes.string,
    overlayClassName: PropTypes.string,
    animation: PropTypes.any,
    align: PropTypes.object,
    overlayStyle: PropTypes.object,
    placement: PropTypes.string,
    overlay: PropTypes.node,
    trigger: PropTypes.array,
    showAction: PropTypes.array,
    hideAction: PropTypes.array,
    getPopupContainer: PropTypes.func,
    visible: PropTypes.bool,
    defaultVisible: PropTypes.bool,
  };

  static defaultProps = {
    minOverlayWidthMatchTrigger: true,
    prefixCls: 'rc-dropdown',
    trigger: ['hover'],
    showAction: [],
    hideAction: [],
    overlayClassName: '',
    overlayStyle: {},
    defaultVisible: false,
    onVisibleChange() {
    },
    placement: 'bottomLeft',
  }

  constructor(props) {
    super(props);
    if ('visible' in props) {
      this.state = {
        visible: props.visible,
      };
    } else {
      this.state = {
        visible: props.defaultVisible,
      };
    }

    if (props.trigger.indexOf('contextMenu') !== -1) {
      addEventListener(document, 'mousedown', this.onDocumentClick);
      addEventListener(window, 'blur', this.onWindowBlur);
    }
  }

  componentWillReceiveProps({ visible }) {
    if (visible !== undefined) {
      this.setState({
        visible,
      });
    }
  }

  onClick = (e) => {
    const props = this.props;
    const overlayProps = props.overlay.props;
    // do no call onVisibleChange, if you need click to hide, use onClick and control visible
    if (!('visible' in props)) {
      this.setState({
        visible: false,
      });
    }
    if (props.onOverlayClick) {
      props.onOverlayClick(e);
    }
    if (overlayProps.onClick) {
      overlayProps.onClick(e);
    }
  }

  onContextMenu = (e) => {
    e.preventDefault();
    this.setState({
      visible: true,
    });
  }

  onDocumentClick = (e) => {
    // ContextMenu behavior
    if (!contains(this.trigger.getPopupDomNode(), e.target)) {
      this.setState({
        visible: false,
      });
    }
  }

  onWindowBlur = () => {
    // ContextMenu behavior
    this.setState({
      visible: false,
    });
  }

  onVisibleChange = (visible) => {
    const props = this.props;
    if (!('visible' in props)) {
      this.setState({
        visible,
      });
    }
    props.onVisibleChange(visible);
  }

  getMenuElement() {
    const { overlay, prefixCls } = this.props;
    const extraOverlayProps = {
      prefixCls: `${prefixCls}-menu`,
      onClick: this.onClick,
    };
    if (typeof overlay.type === 'string') {
      delete extraOverlayProps.prefixCls;
    }
    return React.cloneElement(overlay, extraOverlayProps);
  }

  getPopupDomNode() {
    return this.trigger.getPopupDomNode();
  }

  afterVisibleChange = (visible) => {
    if (visible && this.props.minOverlayWidthMatchTrigger) {
      const overlayNode = this.getPopupDomNode();
      const rootNode = ReactDOM.findDOMNode(this);
      if (rootNode && overlayNode && rootNode.offsetWidth > overlayNode.offsetWidth) {
        overlayNode.style.width = `${rootNode.offsetWidth}px`;
        if (this.trigger &&
            this.trigger._component &&
            this.trigger._component.alignInstance) {
          this.trigger._component.alignInstance.forceAlign();
        }
      }
    }
  }

  saveTrigger = (node) => {
    this.trigger = node;
  }

  render() {
    const {
      prefixCls, children,
      transitionName, animation,
      align, placement, getPopupContainer,
      showAction, hideAction,
      overlayClassName, overlayStyle,
      trigger, ...otherProps,
    } = this.props;
    return (
      <div onContextMenu={this.onContextMenu}>
        <Trigger
          {...otherProps}
          prefixCls={prefixCls}
          ref={this.saveTrigger}
          popupClassName={overlayClassName}
          popupStyle={overlayStyle}
          builtinPlacements={placements}
          action={trigger}
          showAction={showAction}
          hideAction={hideAction}
          popupPlacement={placement}
          popupAlign={align}
          popupTransitionName={transitionName}
          popupAnimation={animation}
          popupVisible={this.state.visible}
          afterPopupVisibleChange={this.afterVisibleChange}
          popup={this.getMenuElement()}
          onPopupVisibleChange={this.onVisibleChange}
          getPopupContainer={getPopupContainer}
        >
          {children}
        </Trigger>
      </div>
    );
  }
}
