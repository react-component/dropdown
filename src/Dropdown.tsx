import type { TriggerProps } from '@rc-component/trigger';
import Trigger from '@rc-component/trigger';
import type {
  ActionType,
  AlignType,
  AnimationType,
  BuildInPlacements,
} from '@rc-component/trigger/lib/interface';
import classNames from 'classnames';
import { composeRef, supportRef } from 'rc-util/lib/ref';
import type { ReactElement } from 'react';
import React from 'react';
import useAccessibility from './hooks/useAccessibility';
import Overlay from './Overlay';
import Placements from './placements';

export interface DropdownProps
  extends Pick<
    TriggerProps,
    | 'getPopupContainer'
    | 'children'
    | 'mouseEnterDelay'
    | 'mouseLeaveDelay'
    | 'onPopupAlign'
    | 'builtinPlacements'
    | 'autoDestroy'
  > {
  minOverlayWidthMatchTrigger?: boolean;
  arrow?: boolean;
  onVisibleChange?: (visible: boolean) => void;
  onOverlayClick?: (e: Event) => void;
  prefixCls?: string;
  transitionName?: string;
  overlayClassName?: string;
  openClassName?: string;
  animation?: AnimationType;
  align?: AlignType;
  overlayStyle?: React.CSSProperties;
  placement?: keyof typeof Placements;
  placements?: BuildInPlacements;
  overlay?: (() => React.ReactElement) | React.ReactElement;
  trigger?: ActionType | ActionType[];
  alignPoint?: boolean;
  showAction?: ActionType[];
  hideAction?: ActionType[];
  visible?: boolean;
  autoFocus?: boolean;
}

function Dropdown(props: DropdownProps, ref) {
  const {
    arrow = false,
    prefixCls = 'rc-dropdown',
    transitionName,
    animation,
    align,
    placement = 'bottomLeft',
    placements = Placements,
    getPopupContainer,
    showAction,
    hideAction,
    overlayClassName,
    overlayStyle,
    visible,
    trigger = ['hover'],
    autoFocus,
    overlay,
    children,
    onVisibleChange,
    ...otherProps
  } = props;

  const [triggerVisible, setTriggerVisible] = React.useState<boolean>();
  const mergedVisible = 'visible' in props ? visible : triggerVisible;

  const triggerRef = React.useRef(null);
  const overlayRef = React.useRef(null);
  const childRef = React.useRef(null);
  React.useImperativeHandle(ref, () => triggerRef.current);

  const handleVisibleChange = (newVisible: boolean) => {
    setTriggerVisible(newVisible);
    onVisibleChange?.(newVisible);
  };

  useAccessibility({
    visible: mergedVisible,
    triggerRef: childRef,
    onVisibleChange: handleVisibleChange,
    autoFocus,
    overlayRef,
  });

  const onClick = (e) => {
    const { onOverlayClick } = props;
    setTriggerVisible(false);

    if (onOverlayClick) {
      onOverlayClick(e);
    }
  };

  const getMenuElement = () => (
    <Overlay
      ref={overlayRef}
      overlay={overlay}
      prefixCls={prefixCls}
      arrow={arrow}
    />
  );

  const getMenuElementOrLambda = () => {
    if (typeof overlay === 'function') {
      return getMenuElement;
    }
    return getMenuElement();
  };

  const getMinOverlayWidthMatchTrigger = () => {
    const { minOverlayWidthMatchTrigger, alignPoint } = props;
    if ('minOverlayWidthMatchTrigger' in props) {
      return minOverlayWidthMatchTrigger;
    }

    return !alignPoint;
  };

  const getOpenClassName = () => {
    const { openClassName } = props;
    if (openClassName !== undefined) {
      return openClassName;
    }
    return `${prefixCls}-open`;
  };

  const childrenNode = React.cloneElement(children, {
    className: classNames(
      children.props?.className,
      mergedVisible && getOpenClassName(),
    ),
    ref: supportRef(children)
      ? composeRef(
          childRef,
          (children as ReactElement & { ref: React.Ref<HTMLElement> }).ref,
        )
      : undefined,
  });

  let triggerHideAction = hideAction;
  if (!triggerHideAction && trigger.indexOf('contextMenu') !== -1) {
    triggerHideAction = ['click'];
  }

  return (
    <Trigger
      builtinPlacements={placements}
      {...otherProps}
      prefixCls={prefixCls}
      ref={triggerRef}
      popupClassName={classNames(overlayClassName, {
        [`${prefixCls}-show-arrow`]: arrow,
      })}
      popupStyle={overlayStyle}
      action={trigger}
      showAction={showAction}
      hideAction={triggerHideAction}
      popupPlacement={placement}
      popupAlign={align}
      popupTransitionName={transitionName}
      popupAnimation={animation}
      popupVisible={mergedVisible}
      stretch={getMinOverlayWidthMatchTrigger() ? 'minWidth' : ''}
      popup={getMenuElementOrLambda()}
      onPopupVisibleChange={handleVisibleChange}
      onPopupClick={onClick}
      getPopupContainer={getPopupContainer}
    >
      {childrenNode}
    </Trigger>
  );
}

export default React.forwardRef(Dropdown);
