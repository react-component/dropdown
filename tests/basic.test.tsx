/* eslint-disable react/button-has-type,react/no-find-dom-node,react/no-render-return-value,object-shorthand,func-names,max-len */
import { act, fireEvent } from '@testing-library/react';
import type { MenuRef } from 'rc-menu';
import Menu, { Divider, Item as MenuItem } from 'rc-menu';
import { _rs } from 'rc-resize-observer';
import { spyElementPrototypes } from 'rc-util/lib/test/domHook';
import type { HTMLAttributes } from 'react';
import * as React from 'react';
import { createRef, forwardRef, useImperativeHandle } from 'react';
import Dropdown from '../src';
import { render, sleep } from './utils';

// Fix prettier rm this
console.log(!!React);

async function waitForTime() {
  for (let i = 0; i < 10; i += 1) {
    await act(async () => {
      jest.runAllTimers();
    });
  }
}

async function triggerResize(target: Element) {
  act(() => {
    _rs([{ target } as ResizeObserverEntry]);
  });

  await waitForTime();
}

spyElementPrototypes(HTMLElement, {
  offsetParent: {
    get: () => document.body,
  },
  offsetLeft: {
    get: function () {
      return parseFloat(window.getComputedStyle(this).marginLeft) || 0;
    },
  },
  offsetTop: {
    get: function () {
      return parseFloat(window.getComputedStyle(this).marginTop) || 0;
    },
  },
  offsetHeight: {
    get: function () {
      return parseFloat(window.getComputedStyle(this).height) || 0;
    },
  },
  offsetWidth: {
    get: function () {
      return parseFloat(window.getComputedStyle(this).width) || 0;
    },
  },

  getBoundingClientRect: () => ({
    width: 100,
    height: 100,
  }),
});

describe('dropdown', () => {
  beforeEach(() => {
    jest.clearAllTimers();
  });

  it('default visible', () => {
    const { container } = render(
      <Dropdown overlay={<div className="check-for-visible">Test</div>} visible>
        <button className="my-button">open</button>
      </Dropdown>,
    );
    expect(container instanceof HTMLDivElement).toBeTruthy();
    expect(
      container
        .querySelector('.my-button')
        ?.classList.contains('rc-dropdown-open'),
    ).toBeTruthy();
  });

  it('supports controlled visible prop', () => {
    const onVisibleChange = jest.fn();
    const { container } = render(
      <Dropdown
        overlay={<div className="check-for-visible">Test</div>}
        visible
        trigger={['click']}
        onVisibleChange={onVisibleChange}
      >
        <button className="my-button">open</button>
      </Dropdown>,
    );
    expect(container instanceof HTMLDivElement).toBeTruthy();
    expect(
      container
        .querySelector('.my-button')
        ?.classList.contains('rc-dropdown-open'),
    ).toBeTruthy();

    fireEvent.click(container.querySelector('.my-button'));
    expect(onVisibleChange).toHaveBeenCalledWith(false);
  });

  it('simply works', async () => {
    let clicked;

    function onClick({ key }) {
      clicked = key;
    }

    const onOverlayClick = jest.fn();

    const menu = (
      <Menu style={{ width: 140 }} onClick={onClick}>
        <MenuItem key="1">
          <span className="my-menuitem">one</span>
        </MenuItem>
        <Divider />
        <MenuItem key="2">two</MenuItem>
      </Menu>
    );
    const { container, baseElement } = render(
      <Dropdown
        trigger={['click']}
        overlay={menu}
        onOverlayClick={onOverlayClick}
      >
        <button className="my-button">open</button>
      </Dropdown>,
    );
    expect(container.querySelector('.my-button')).toBeTruthy();
    // should not display until be triggered
    expect(baseElement.querySelector('.rc-dropdown')).toBeFalsy();

    fireEvent.click(container.querySelector('.my-button'));
    expect(clicked).toBeUndefined();
    expect(
      baseElement
        .querySelector('.rc-dropdown')
        .classList.contains('rc-dropdown-hidden'),
    ).toBeFalsy();
    expect(container).toMatchSnapshot();

    fireEvent.click(baseElement.querySelector('.my-menuitem'));
    expect(clicked).toBe('1');
    expect(onOverlayClick).toHaveBeenCalled();
    expect(
      baseElement
        .querySelector('.rc-dropdown')
        .classList.contains('rc-dropdown-hidden'),
    ).toBeTruthy();
  });

  it('re-align works', async () => {
    jest.useFakeTimers();

    const onPopupAlign = jest.fn();

    const buttonStyle = { width: 600, height: 20, marginLeft: 100 };
    const menu = (
      <Menu>
        <MenuItem key="1">one</MenuItem>
      </Menu>
    );
    const { container } = render(
      <Dropdown
        trigger={['click']}
        placement="bottomRight"
        overlay={menu}
        onPopupAlign={onPopupAlign}
      >
        <button className="my-btn" style={buttonStyle}>
          open
        </button>
      </Dropdown>,
    );

    expect(onPopupAlign).not.toHaveBeenCalled();

    fireEvent.click(container.querySelector('.my-btn'));
    await waitForTime();

    expect(onPopupAlign).toHaveBeenCalled();

    jest.useRealTimers();
  });

  it('Test default minOverlayWidthMatchTrigger', async () => {
    jest.useFakeTimers();

    const overlayWidth = 50;
    const overlay = <div style={{ width: overlayWidth }}>Test</div>;

    const { container, baseElement } = render(
      <Dropdown trigger={['click']} overlay={overlay} visible>
        <button style={{ width: 100 }} className="my-button">
          open
        </button>
      </Dropdown>,
    );

    await triggerResize(container.querySelector('button'));

    expect(baseElement.querySelector('.rc-dropdown')).toHaveStyle({
      minWidth: '100px',
    });

    jest.useRealTimers();
  });

  it('user pass minOverlayWidthMatchTrigger', async () => {
    jest.useFakeTimers();

    const overlayWidth = 50;
    const overlay = <div style={{ width: overlayWidth }}>Test</div>;

    const { container, baseElement } = render(
      <Dropdown
        trigger={['click']}
        overlay={overlay}
        minOverlayWidthMatchTrigger={false}
        visible
      >
        <button style={{ width: 100 }} className="my-button">
          open
        </button>
      </Dropdown>,
    );

    await triggerResize(container.querySelector('button'));

    expect(baseElement.querySelector('.rc-dropdown')).not.toHaveStyle({
      minWidth: '100px',
    });

    jest.useRealTimers();
  });

  it('should support default openClassName', () => {
    const overlay = <div style={{ width: 50 }}>Test</div>;
    const { container } = render(
      <Dropdown
        trigger={['click']}
        overlay={overlay}
        minOverlayWidthMatchTrigger={false}
      >
        <button style={{ width: 100 }} className="my-button">
          open
        </button>
      </Dropdown>,
    );
    fireEvent.click(container.querySelector('.my-button'));
    expect(
      container
        .querySelector('.my-button')
        .classList.contains('rc-dropdown-open'),
    ).toBeTruthy();
    fireEvent.click(container.querySelector('.my-button'));
    expect(
      container
        .querySelector('.my-button')
        .classList.contains('rc-dropdown-open'),
    ).toBeFalsy();
  });

  it('should support custom openClassName', async () => {
    const overlay = <div style={{ width: 50 }}>Test</div>;
    const { container } = render(
      <Dropdown
        trigger={['click']}
        overlay={overlay}
        minOverlayWidthMatchTrigger={false}
        openClassName="opened"
      >
        <button style={{ width: 100 }} className="my-button">
          open
        </button>
      </Dropdown>,
    );

    fireEvent.click(container.querySelector('.my-button'));
    expect(
      container.querySelector('.my-button').classList.contains('opened'),
    ).toBeTruthy();
    fireEvent.click(container.querySelector('.my-button'));
    expect(
      container.querySelector('.my-button').classList.contains('opened'),
    ).toBeFalsy();
  });

  it('overlay callback', async () => {
    const overlay = <div style={{ width: 50 }}>Test</div>;
    const { container, baseElement } = render(
      <Dropdown trigger={['click']} overlay={() => overlay}>
        <button className="my-button">open</button>
      </Dropdown>,
    );

    fireEvent.click(container.querySelector('.my-button'));
    expect(
      baseElement
        .querySelector('.rc-dropdown')
        .classList.contains('rc-dropdown-hidden'),
    ).toBeFalsy();
  });

  it('should support arrow', async () => {
    const overlay = <div style={{ width: 50 }}>Test</div>;
    const { container, baseElement } = render(
      <Dropdown arrow overlay={overlay} trigger={['click']}>
        <button style={{ width: 100 }} className="my-button">
          open
        </button>
      </Dropdown>,
    );

    fireEvent.click(container.querySelector('.my-button'));
    await sleep(500);
    expect(
      baseElement
        .querySelector('.rc-dropdown')
        .classList.contains('rc-dropdown-show-arrow'),
    ).toBeTruthy();
    expect(
      baseElement
        .querySelector('.rc-dropdown')
        .firstElementChild.classList.contains('rc-dropdown-arrow'),
    ).toBeTruthy();
  });

  it('Keyboard navigation works', async () => {
    jest.useFakeTimers();

    const overlay = (
      <Menu>
        <MenuItem key="1">
          <span className="my-menuitem">one</span>
        </MenuItem>
        <MenuItem key="2">two</MenuItem>
      </Menu>
    );
    const { container, baseElement } = render(
      <Dropdown trigger={['click']} overlay={overlay}>
        <button className="my-button">open</button>
      </Dropdown>,
    );
    const trigger = container.querySelector('.my-button');

    // Open menu;
    fireEvent.click(trigger);
    await waitForTime();
    expect(
      baseElement
        .querySelector('.rc-dropdown')
        .classList.contains('rc-dropdown-hidden'),
    ).toBeFalsy();

    // Close menu with Esc
    fireEvent.keyDown(window, { key: 'Esc', keyCode: 27 });
    await waitForTime();
    expect(document.activeElement.className).toContain('my-button');

    // Open menu
    fireEvent.click(trigger);
    await waitForTime();
    expect(
      baseElement
        .querySelector('.rc-dropdown')
        .classList.contains('rc-dropdown-hidden'),
    ).toBeFalsy();

    // Focus menu with Tab
    window.dispatchEvent(new KeyboardEvent('keydown', { keyCode: 9 })); // Tab
    expect(document.activeElement.className).toContain('menu');

    // Close menu with Tab
    window.dispatchEvent(new KeyboardEvent('keydown', { keyCode: 9 })); // Tab
    await waitForTime();
    expect(document.activeElement.className).toContain('my-button');

    jest.useRealTimers();
  });

  it('Tab should close menu if overlay cannot be focused', async () => {
    jest.useFakeTimers();

    const Overlay = () => <div>test</div>;
    const { container, baseElement } = render(
      <Dropdown trigger={['click']} overlay={<Overlay />}>
        <button className="my-button">open</button>
      </Dropdown>,
    );
    const trigger = container.querySelector('.my-button');

    // Open menu;
    fireEvent.click(trigger);
    await waitForTime();
    expect(
      baseElement
        .querySelector('.rc-dropdown')
        .classList.contains('rc-dropdown-hidden'),
    ).toBeFalsy();

    // Close menu with Esc
    fireEvent.keyDown(window, { key: 'Esc', keyCode: 27 });
    await waitForTime();
    expect(document.activeElement.className).toContain('my-button');

    // Open menu
    fireEvent.click(trigger);
    await waitForTime();
    expect(
      baseElement
        .querySelector('.rc-dropdown')
        .classList.contains('rc-dropdown-hidden'),
    ).toBeFalsy();

    // Close menu with Tab
    window.dispatchEvent(new KeyboardEvent('keydown', { keyCode: 9 })); // Tab
    await waitForTime();
    expect(document.activeElement.className).toContain('my-button');

    jest.useRealTimers();
  });

  it('keyboard should work if menu is wrapped', async () => {
    const overlay = (
      <div>
        <Menu>
          <MenuItem key="1">
            <span className="my-menuitem">one</span>
          </MenuItem>
          <MenuItem key="2">two</MenuItem>
        </Menu>
      </div>
    );
    const { container, baseElement } = render(
      <Dropdown trigger={['click']} overlay={overlay}>
        <button className="my-button">open</button>
      </Dropdown>,
    );
    const trigger = container.querySelector('.my-button');

    // Open menu
    fireEvent.click(trigger);
    await sleep(200);
    expect(
      baseElement
        .querySelector('.rc-dropdown')
        .classList.contains('rc-dropdown-hidden'),
    ).toBeFalsy();

    // Close menu with Esc
    window.dispatchEvent(new KeyboardEvent('keydown', { keyCode: 27 })); // Esc
    await sleep(200);
    expect(document.activeElement.className).toContain('my-button');

    // Open menu
    fireEvent.click(trigger);
    await sleep(200);
    expect(
      baseElement
        .querySelector('.rc-dropdown')
        .classList.contains('rc-dropdown-hidden'),
    ).toBeFalsy();

    // Focus menu with Tab
    window.dispatchEvent(new KeyboardEvent('keydown', { keyCode: 9 })); // Tab

    // Close menu with Tab
    window.dispatchEvent(new KeyboardEvent('keydown', { keyCode: 9 })); // Tab
    await sleep(200);
    expect(document.activeElement.className).toContain('my-button');
  });

  it('support Menu expandIcon', async () => {
    const props = {
      overlay: (
        <Menu expandIcon={<span id="customExpandIcon" />}>
          <Menu.Item key="1">foo</Menu.Item>
          <Menu.SubMenu title="SubMenu">
            <Menu.Item key="1">foo</Menu.Item>
          </Menu.SubMenu>
        </Menu>
      ),
      visible: true,
      getPopupContainer: (node) => node,
    };

    const { container } = render(
      <Dropdown {...props}>
        <button type="button">button</button>
      </Dropdown>,
    );
    await sleep(500);
    expect(container.querySelector('#customExpandIcon')).toBeTruthy();
  });

  it('should support customized menuRef', async () => {
    const menuRef = createRef<MenuRef>();
    const props = {
      overlay: (
        <Menu ref={menuRef}>
          <Menu.Item key="1">foo</Menu.Item>
        </Menu>
      ),
      visible: true,
    };

    render(
      <Dropdown {...props}>
        <button type="button">button</button>
      </Dropdown>,
    );

    await sleep(500);
    expect(menuRef.current).toBeTruthy();
  });

  it('should support trigger which not support focus', async () => {
    jest.useFakeTimers();
    const Button = forwardRef<any, HTMLAttributes<HTMLButtonElement>>(
      (props, ref) => {
        useImperativeHandle(ref, () => ({
          foo: () => {},
        }));
        return (
          <button
            onClick={(e) => {
              props?.onClick?.(e);
            }}
          >
            trigger
          </button>
        );
      },
    );
    const { container, baseElement } = render(
      <Dropdown
        trigger={['click']}
        getPopupContainer={(node) => node}
        overlay={
          <Menu>
            <Menu.Item key="1">foo</Menu.Item>
          </Menu>
        }
      >
        <Button />
      </Dropdown>,
    );
    fireEvent.click(container.querySelector('button'));
    fireEvent.click(baseElement.querySelectorAll('li')[0]);

    jest.runAllTimers();
    jest.useRealTimers();
  });

  it('should support autoFocus', async () => {
    jest.useFakeTimers();

    const overlay = (
      <Menu>
        <MenuItem key="1">
          <span className="my-menuitem">one</span>
        </MenuItem>
        <MenuItem key="2">two</MenuItem>
      </Menu>
    );
    const { container } = render(
      <Dropdown autoFocus trigger={['click']} overlay={overlay}>
        <button className="my-button">open</button>
      </Dropdown>,
    );
    const trigger = container.querySelector('.my-button');

    // Open menu
    fireEvent.click(trigger);

    await waitForTime();

    expect(
      container
        .querySelector('.rc-dropdown')
        .classList.contains('rc-dropdown-hidden'),
    ).toBeFalsy();
    expect(document.activeElement.className).toContain('menu');

    // Close menu with Tab
    window.dispatchEvent(new KeyboardEvent('keydown', { keyCode: 9 })); // Tab

    await waitForTime();

    expect(document.activeElement.className).toContain('my-button');

    jest.useRealTimers();
  });

  it('children cannot be given ref should not throw', () => {
    const errorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});
    const Component = () => <div>test</div>;

    render(
      <Dropdown overlay={<div>test</div>}>
        <Component />
      </Dropdown>,
    );
    expect(errorSpy).not.toHaveBeenCalledWith(
      expect.stringContaining(
        'Warning: Function components cannot be given refs',
      ),
      expect.anything(),
      expect.anything(),
    );
  });
});
