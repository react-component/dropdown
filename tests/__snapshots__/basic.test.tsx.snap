// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dropdown simply works 1`] = `
<div>
  <button
    class="my-button rc-dropdown-open"
  >
    open
  </button>
  <div
    class="rc-dropdown rc-dropdown-placement-bottomLeft"
    style="--arrow-x: 0px; --arrow-y: 0px; left: -1000vw; top: -1000vh; box-sizing: border-box; min-width: 100px;"
  >
    <ul
      class="rc-menu rc-menu-root rc-menu-vertical"
      data-menu-list="true"
      role="menu"
      style="width: 140px;"
      tabindex="0"
    >
      <li
        class="rc-menu-item"
        data-menu-id="rc-menu-uuid-test-1"
        role="menuitem"
        tabindex="-1"
      >
        <span
          class="my-menuitem"
        >
          one
        </span>
      </li>
      <li
        class="rc-menu-item-divider"
        role="separator"
      />
      <li
        class="rc-menu-item"
        data-menu-id="rc-menu-uuid-test-2"
        role="menuitem"
        tabindex="-1"
      >
        two
      </li>
    </ul>
    <div
      aria-hidden="true"
      style="display: none;"
    />
  </div>
</div>
`;
